### PR TITLE
Fix underscore in binary, Int and hex numbers

### PR DESCRIPTION
**Pull request by Armanbqt**
This is to fix invalid placement of underscore in numbers:

- Multiple underscores: ```12___3```
- AfterRadixPrefix: ```0x_76```
- BeforeDecimalPoint: ```12_.44```
- Trailing underscores: ```0b1011_```

### CODE REVIEW

File: ion/tokenizer.go [hunk 1]
```diff
@@ -755,6 +755,14 @@ func (t *tokenizer) readRadix(pok, dok matcher) (string, error) {
 	}
 	w.WriteByte(byte(c))
 
+	// At this point we have either 0x or 0b, and it cannot be followed by _
+	nextChar, err2 := t.peek()
+	if err2 != nil {
+		return "", err
+	}
+	if nextChar == '_' {
```
Comment (zslayton) [inline ion/tokenizer.go:763]: Would it make more sense for this `_` check to live inside `readRadixDigits` [just before the `for` loop](https://github.com/amzn/ion-go/blob/74c4cf0a310ebbab78a1b844b055fb08e6aa953c/ion/tokenizer.go#L787) so other calls benefit from the check too?

File: ion/tokenizer.go [hunk 2]
```diff
@@ -782,12 +790,22 @@ func (t *tokenizer) readRadixDigits(dok matcher, w io.ByteWriter) (int, error) {
 			return 0, err
 		}
 		if c == '_' {
+			nextChar, err := t.peek()
+			if err != nil {
+				return 0, err
+			}
+			if !dok(nextChar) {
```
[the remark this answers is not in the document]
Comment (Armanbqt) [reply ion/tokenizer.go:797]: I agree that the names are not very descriptive, however, we cannot rename them to ```isX``` and ```isHexDigit```. 
Depending on the number, ```readRadix``` might be called by either [```readHex()```](https://github.com/amzn/ion-go/blob/74c4cf0a310ebbab78a1b844b055fb08e6aa953c/ion/tokenizer.go#L711) or [```readBinary()```](https://github.com/amzn/ion-go/blob/74c4cf0a310ebbab78a1b844b055fb08e6aa953c/ion/tokenizer.go#L721) which for the latter case ```pok``` and ```dok``` are going to be ```isB``` and ```isDigit``` respectively.

How about ```isInBase``` and ```isValidInBase``` - not the best names, but cannot think of anything better!

File: ion/tokenizer.go (hunk 1 above)
Comment (Armanbqt) [reply ion/tokenizer.go:763]: We cannot. [```readDigits()```](https://github.com/amzn/ion-go/blob/74c4cf0a310ebbab78a1b844b055fb08e6aa953c/ion/tokenizer.go#L446) also calls ```readRadixDigits``` function and if we add this ```_``` check in ```readRadixDigits```, then a valid number like ```1_2.3_4``` will be considered as invalid (first digit will be taken in ```readDigits()``` and then when we get to ```readRadixDigits``` the first character is ```_```)

Comment (zslayton) [reply ion/tokenizer.go:763]: Got it, thanks!

File: ion/tokenizer.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (zslayton) [reply ion/tokenizer.go:797]: How about `isRadixMarker` for detecting `x`/`b` and `isValidForRadix` for checking each digit?

[the remark this answers is not in the document]
Comment (Armanbqt) [reply ion/tokenizer.go]: Renamed. Thanks.

**Comment by zslayton** [comment]
Looks good. 👍 Could you merge from `master` and resolve the conflicts in a new commit? Then I can squash+merge.

**Pull request merged**